go-controller/pkg/ovn/controller/egressfirewall/egressfirewall.go

```diff
@@ -188,120 +289,104 @@ func (oc *DefaultNetworkController) syncEgressFirewall(egressFirewalls []interfa
-	if err != nil {
-		return fmt.Errorf("error deleting egress firewall policies on router %s: %v", oc.GetNetworkScopedClusterRouterName(), err)
-	}
+func (oc *EFController) Start() error {
+	klog.Infof("Starting EgressFirewall controller")
+	return controller.StartWithInitialSync(oc.initialSync, oc.controller, oc.nodeController)
+}
 
-	// delete acls from all switches, they reside on the port group now
-	// Lookup all ACLs used for egress Firewalls
-	aclPred := func(item *nbdb.ACL) bool {
-		return item.Priority >= types.MinimumReservedEgressFirewallPriority && item.Priority <= types.EgressFirewallStartPriority
-	}
-	egressFirewallACLs, err := libovsdbops.FindACLsWithPredicate(oc.nbClient, aclPred)
+func (oc *EFController) Stop() {
+	klog.Infof("%s: shutting down", oc.name)
+	controller.Stop(oc.nodeController, oc.controller)
+}
+
+func (oc *EFController) sync(key string) (updateErr error) {
+	oc.RLock()
+	defer oc.RUnlock()
+	klog.V(5).Info("Syncing EgressFirewall", key)
+	namespace, efName, err := cache.SplitMetaNamespaceKey(key)
 	if err != nil {
-		return fmt.Errorf("unable to list egress firewall ACLs, cannot cleanup old stale data, err: %v", err)
+		return fmt.Errorf("invalid resource key for egress firewall: %s", key)
 	}
-	if len(egressFirewallACLs) != 0 {
-		err = batching.Batch[*nbdb.ACL](aclDeleteBatchSize, egressFirewallACLs, func(batchACLs []*nbdb.ACL) error {
-			// optimize the predicate to exclude switches that don't reference deleting acls.
-			aclsToDelete := sets.NewString()
-			for _, acl := range batchACLs {
-				aclsToDelete.Insert(acl.UUID)
-			}
-			swWithACLsPred := func(sw *nbdb.LogicalSwitch) bool {
-				return aclsToDelete.HasAny(sw.ACLs...)
-			}
-			return libovsdbops.RemoveACLsFromLogicalSwitchesWithPredicate(oc.nbClient, swWithACLsPred, batchACLs...)
-		})
+	ef, err := oc.efLister.EgressFirewalls(namespace).Get(efName)
+	if err != nil {
+		if !apierrors.IsNotFound(err) {
+			return err
+		}
+		// delete case
+		klog.V(3).Infof("Removing egress firewall %s", key)
+		p := libovsdbops.GetPredicate[*nbdb.ACL](oc.GetEgressFirewallACLDbIDsNoRule(namespace), nil)
+		invalidACLs, err := libovsdbops.FindACLsWithPredicate(oc.nbClient, p)
 		if err != nil {
-			return fmt.Errorf("failed to remove egress firewall acls from node logical switches: %v", err)
+			return fmt.Errorf("error finding ACLs for egress firewall %s: %w", key, err)
+		}
+		if err := libovsdbops.DeleteACLsFromAllPortGroups(oc.nbClient, invalidACLs...); err != nil {
+			return fmt.Errorf("error deleting stale ACLs for egress firewall %s: %w", key, err)
+		}
+		if err := oc.dnsNameResolver.Delete(namespace); err != nil {
+			return err
 		}
+		numRules, ok := oc.ruleCounter.Load(key)
+		if ok {
+			metrics.UpdateEgressFirewallRuleCount(float64(-numRules.(uint32)))
+			oc.ruleCounter.Delete(key)
+		} else {
+			klog.Errorf("Unable to decrement egress firewall rule count, cache miss for key: %s", key)
+		}
+		metrics.DecrementEgressFirewallCount()
```

Comment:
I find it confusing that metrics are handled here for delete, but in `setEgressFirewallStatus` for update. Maybe we can update `setEgressFirewallStatus` to handle both and just tell it if was update or delete?